### PR TITLE
fix : lookup word in vocab before doing BPE merges

### PR DESCRIPTION
**Pull request by tonyfettes**
For llama-3, I found there is an inconsistency between llama.cpp's tokenizer and [Huggingface's tokenizers](https://github.com/huggingface/tokenizers). Example:

```
 Việt
```

llama.cpp:
```
 11655 -> ' Vi'
 26298 -> 'ệ'
    83 -> 't'
```

Huggingface's tokenizers with tokenizer.json from llama-3:
```
101798
```

After comparing the implementation, it seems that Huggingface's tokenizers will try to lookup a split word in the vocabulary first, and push to the result tokens if found; if not, it will try to merge the word at byte level instead. In llama.cpp, we always do the byte-level merge, hence the inconsistency.

This is a simple fix to the problem, by just looking the word up before do the merging.

PS: I have checked with tiktoken and it seems they did the same thing at `src/lib.rs:228` in `CoreBPE::_encode_native`

PPS: I searched `tokenizer.json` from all BPE models (some are license-walled so I checked their variants) and it seems that llama-3 is the only one doing this?

|Model|tokenizer.json|
|---|---|
|DBRX (Walled)|https://huggingface.co/turboderp/dbrx-instruct-exl2/tree/2.3bpw|
|Deepseek LLM|https://huggingface.co/deepseek-ai/deepseek-llm-67b-chat/raw/main/tokenizer.json|
|Deepseek Coder|https://huggingface.co/deepseek-ai/deepseek-coder-33b-instruct/raw/main/tokenizer.json|
|Falcon|https://huggingface.co/tiiuae/falcon-7b/raw/main/tokenizer.json|
|Starcoder|https://huggingface.co/bigcode/starcoder/raw/main/tokenizer.json|
|Refact|https://huggingface.co/smallcloudai/Refact-1_6B-fim/blob/main/tokenizer.json|
|Command R+|https://huggingface.co/CohereForAI/c4ai-command-r-plus/blob/main/tokenizer.json|
|GPT2|https://huggingface.co/openai-community/gpt2/raw/main/tokenizer.json|
|OLMo|https://huggingface.co/allenai/OLMo-7B-Instruct/raw/main/tokenizer.json|
|Qwen2 (Qwen1.5)|https://huggingface.co/Qwen/Qwen1.5-110B-Chat/raw/main/tokenizer.json|

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -12200,13 +12200,14 @@ struct llm_tokenizer_bpe {
 
     void tokenize(const std::string & text, std::vector<llama_vocab::id> & output) {
         int final_prev_index = -1;
+        int ignore_merges = false;
```
Comment (ggerganov) [inline llama.cpp:12203]:
```suggestion
        bool ignore_merges = false;
```

**Comment by ggerganov** [comment]
> PPS: I searched tokenizer.json from all BPE models (some are license-walled so I checked their variants) and it seems that llama-3 is the only one doing this?

Which parameter in the tokenizer config determines this behaviour?

**Comment by tonyfettes** [comment]
@ggerganov `"ignore_merges"`, under `"model"`

**Review by ggerganov — APPROVED** [review]
Let's merge after the green is CI

File: llama.cpp [hunk 2]
```diff
@@ -12292,6 +12299,20 @@ struct llm_tokenizer_bpe {
...
+                sym.prev = final_prev_index;
+                sym.next = -1;
+                if (final_prev_index != -1) {
+                    symbols_final[final_prev_index].next = symbols_final.size();
+                }
+                symbols_final.emplace_back(sym);
+                final_prev_index = symbols_final.size() - 1;
+                continue;
+            }
+
```
Comment (ggerganov) [inline llama.cpp:12315]: Let's apply @jaime-m-p's suggestion here, to reduce the code duplication in this loop:

https://github.com/ggerganov/llama.cpp/pull/6965#issuecomment-2105495718

**Pull request merged**